### PR TITLE
fix(json): reach full JSON test262 conformance

### LINKED ISSUE #845: Bring JSON to 100% test262 conformance (~34 fails)
## Summary

Drive the test262 **JSON** failures to zero across `built-ins`, `staging`, and `intl402` — these share root causes, so a single fix typically clears the failure in more than one category.

## Why

Reaching ~100% test262 on supported categories is the objective for the conformance milestone. Baseline (main, 2026-06-26): **97.3% overall** (50,877/52,293; 0 timeouts, 0 infra; `language` and `harness` already at 100%). **JSON accounts for ~34** of the 1,416 remaining failures.

## Current behavior

~34 JSON tests fail. Concentrations: JSON.parse / stringify edge cases and source-text access.

## Expected behavior

Zero JSON failures at the pinned suite SHA (`scripts/test262-suite-sha.txt`), with no regressions elsewhere — the PR conformance gate enforces no PASS to non-PASS transitions.

## Scope notes

- Organize the work by failing sub-area (above); they are independently fixable.
- Land the conformance gates first: harness diagnostics (#830), the crash fix (#831), the skip-list (#832).
- Every fix ships with its JS test in the same commit. Parent #820.

### PR DESCRIPTION
**Pull request by frostney**
## Summary

- Bring JSON.parse/JSON.stringify behavior in line with ES2026 for parse records, reviver source contexts, proxy-aware JSON arrays, function arity, missing-argument coercion, BigInt toJSON ordering, and QuoteJSONString escaping.
- Run native callbacks, native constructors, and Object() primitive boxing through the correct creation realm so cross-realm JSON.rawJSON errors and boxed BigInt toJSON lookup use the expected realm.
- Clean up review findings for aborted JSON parse record collection and native Pascal JSON escaping assertions.
- Closes #845.

## Testing

- [x] Verified no regressions and confirmed the new feature or bugfix in end-to-end JavaScript/TypeScript tests
- [ ] Updated documentation
- [x] Optional: Verified no regressions and confirmed the new feature or bugfix in native Pascal tests (if AST, scope, evaluator, or value types changed)
- [ ] Optional: Verified no benchmark regressions or confirmed benchmark coverage for the change

Validation run:
- `./build.pas --clean loaderbare`
- `./build.pas --clean testrunner`
- `./build.pas testrunner`
- `./build/GocciaTestRunner tests/built-ins/JSON/parse.js`
- `./build/GocciaTestRunner tests/built-ins/JSON/stringify.js`
- `./build/GocciaTestRunner tests/built-ins/JSON/parse.js --mode=bytecode`
- `./build/GocciaTestRunner tests/built-ins/JSON/stringify.js --mode=bytecode`
- `./build/GocciaTestRunner tests`
- `./build/GocciaTestRunner tests --mode=bytecode`
- `mkdir -p build/compiled/tests/source/units/Goccia.JSON.Utils.Test && fpc @config.cfg -FUbuild/compiled/tests/source/units/Goccia.JSON.Utils.Test -O- -gw -godwarfsets -gl -Ct -Cr -Sa -vw-n-h-i-l-d-u-t-p-c-x- source/units/Goccia.JSON.Utils.Test.pas && ./build/Goccia.JSON.Utils.Test`
- `./build/Goccia.Engine.Realm.Test`
- `./build/GocciaTestRunner tests/built-ins/JSON/parse.js --no-progress --silent && ./build/GocciaTestRunner tests/built-ins/JSON/parse.js --mode=bytecode --no-progress --silent`
- `./build.pas tests`
- `for t in build/Goccia.*.Test(N); do "$t"; done`
- `./build.pas testrunner && ./build/GocciaTestRunner tests --no-progress --silent && ./build/GocciaTestRunner tests --mode=bytecode --no-progress --silent`
- `./format.pas --check`
- `git diff --check`
- pinned test262 bytecode `**/JSON/**/*.js`: 159/159 passed
- pinned test262 bytecode `staging/sm/JSON/*.js`: 39/39 passed

### CODE REVIEW

File: source/units/Goccia.JSON.pas [hunk 1]
```diff
@@ -313,14 +438,19 @@ constructor TGocciaJSONVisitor.Create(
...
   FResult := nil;
+  FRootRecord := nil;
+  FCollectRecords := False;
   FCollectSources := False;
 end;
 
 destructor TGocciaJSONVisitor.Destroy;
 begin
+  FRootRecord.Free;
+  FRecordStack.Free;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.JSON.pas:458]: Fixed in f0d4842c. TGocciaJSONVisitor now frees any parse records still left on FRecordStack when parsing aborts before ownership transfers to the root/parent record.

File: source/units/Goccia.Values.NativeFunction.pas [hunk 2]
```diff
@@ -72,8 +73,18 @@ constructor TGocciaNativeFunctionValue.CreateWithoutPrototype(const AFunction: T
...
+  PreviousRealm := CurrentRealm;
+  if Assigned(FCreationRealm) and (FCreationRealm <> PreviousRealm) then
+    SetCurrentRealm(FCreationRealm);
+  try
+    Result := FFunction(AArguments, AThisValue);
+  finally
+    if Assigned(FCreationRealm) and (FCreationRealm <> PreviousRealm) then
+      SetCurrentRealm(PreviousRealm);
+  end;
 end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Values.NativeFunction.pas:88]: Fixed in f0d4842c. Native function construction now switches to the function creation realm and restores the caller realm in a finally block, matching Call for both construct callbacks and the fallback FFunction path.
